### ISSUE TITLE
Job to schedule SMSs on Infobip based on updates in CommCare

### ISSUE THREAD

**Issue by ritazagoni** [OpenFn/Miracle-Feet#27]
## Background, context, and business value

Currently MiracleFeet uses OpenFn to integrate CommCare case updates and form submissions with Salesforce and CommCare to send SMSs to patients. They seek to lower their SMS transactions costs by expanding the existing OpenFn integration to also schedule SMSs in Infobip whenever a CommCare update is received.

- As a case worker, when I create or update a patient record in CommCare, I want to have an SMS or a series of SMSs automatically scheduled to be sent to the patient.
- As a case worker, when I create or update a visit record in CommCare, I want to have an SMS or a series of SMSs automatically scheduled to be sent to the patient.

## The specific request, in as few words as possible

Create a job that executes the following steps (see Actions section for more detail)...
1. CommCare will send OpenFn a notification with triggering sms information (see state)
2. OpenFn should set a `bulkId` for all relevant message types (see mapping specs for instructions)
3. Then use these `bulkId`-s to query the Infobip API to fetch which messages have already been scheduled via `GET '/sms/1/bulks'`. 
4. Using `bulkId` to match and compare the response to step #3 to determine which SMSs already exist in Infobip...
5a. If SMS does _not_ exist in Infobip, schedule new SMS via `POST '/sms/2/text/advanced'`
5b. If SMS does exist in Infobip, reschedule the SMS using `bulkId`  via `PUT '/sms/1/bulks'`

Note that we're replicating an upsert pattern here where we first check for existing SMSs, and then schedule or reschedule depending on whether that SMS already exists. 

**Design docs:** 
1. See [data flow](https://lucid.app/invitations/accept/b65ed007-8a44-48ac-8ee7-38075c45e460)
2. See [mapping specifications](https://docs.google.com/spreadsheets/d/1quhQJgQkVRC8oObDzkwgnnm-Rov5BGOW85I4YqcNV0I/edit?usp=sharing)
3. See list of [alerts and corresponding triggers](https://docs.google.com/spreadsheets/d/1quhQJgQkVRC8oObDzkwgnnm-Rov5BGOW85I4YqcNV0I/edit#gid=542770724)
4. See [message content in different languages](https://docs.google.com/spreadsheets/d/1quhQJgQkVRC8oObDzkwgnnm-Rov5BGOW85I4YqcNV0I/edit#gid=1959335211)
5. See this [sample job](https://gist.github.com/aleksa-krolls/6af43936d55d68383d5c0dce27d84d64) that defines message templates in different languages and posts an HTTP request to send a message through Infobip with a template selected based on properties in state. It also assigns a bulkId to each message. Issue will be updated with exactly how to construct the bulkId.
6. Use API docs: 
https://www.infobip.com/docs/api#programmable-communications/sms/send-sms-message

### Trigger
This job will run when `"sms_opt_in": "yes"` --> the job will then need to evaluate different CommCare case properties to determine which SMS type(s) should be scheduled. See the `Triggering Criteria` column in the [Alerts mapping specs](https://docs.google.com/spreadsheets/d/1quhQJgQkVRC8oObDzkwgnnm-Rov5BGOW85I4YqcNV0I/edit#gid=542770724). 

This job may trigger 1 or multiple SMSs depending on what properties are included in `state`. The case `language_code` will then determine which language these SMSs will be sent. 

### Actions
The jobs should automate the following actions:
1. Construct unique bulkId for every relevant message for a case based on properties in `state`.
2. Query scheduled SMSs for a case on Infobip by `bulkId`.
API docs: https://www.infobip.com/docs/api#programmable-communications/sms/get-scheduled-sms-messages
3. Determine which message(s) to schedule based on `bulkId` and `language_code` based on whether message already have been scheduled in Infobip (see step 2).
4. If message hasn't been scheduled yet, schedule SMS(s) on Infobip.
API docs: https://www.infobip.com/docs/api#programmable-communications/sms/send-sms-message
5. If message has been already scheduled, reschedule the message for the same date and time
API docs: https://www.infobip.com/docs/api#programmable-communications/sms/reschedule-sms-messages

You can see message logs on Infobip at https://portal.infobip.com/analyze/logs.

## state.json
See these two example payloads from CommCare.
1. https://github.com/OpenFn/Miracle-Feet/blob/master/sample_data/patientRegistration_educationalMessage.json
2. https://github.com/OpenFn/Miracle-Feet/blob/master/sample_data/newVisit_ReminderMessage.json

For Infobip login see LP `Infobip Portal (MiracleFeet)`. 

## adaptor
`language-http`

## expression.js
Create 1 new job `scheduleMessages.js`

## output.json
1. Every time a case is created or updated in CommCare, determine which message(s) need to be scheduled. Schedule new messages and reschedule existing messages.

For the sample state for [patient registration](https://github.com/OpenFn/Miracle-Feet/blob/master/sample_data/patientRegistration_educationalMessage.json) we have the following potential triggers in `state` (see potential triggers [here](https://docs.google.com/spreadsheets/d/1quhQJgQkVRC8oObDzkwgnnm-Rov5BGOW85I4YqcNV0I/edit#gid=2026279226):

`"send_sms": "on"`
`"sms_opt_in": "yes"`
`"treatment": "bracing_night"`

Based on [alerts and their triggers](https://docs.google.com/spreadsheets/d/1quhQJgQkVRC8oObDzkwgnnm-Rov5BGOW85I4YqcNV0I/edit#gid=542770724), messages for `alerts` `8`, `9`, `10`, `11` and `12` need to be scheduled according to [Schedule start date, Schedule date, Schedule time](https://docs.google.com/spreadsheets/d/1quhQJgQkVRC8oObDzkwgnnm-Rov5BGOW85I4YqcNV0I/edit#gid=542770724&range=F:H).

Also from `state`:
`"language_code": ""`
the language of the messages has to be the default English.

The exact messages that will be sent out in this scenario are [here](https://docs.google.com/spreadsheets/d/1quhQJgQkVRC8oObDzkwgnnm-Rov5BGOW85I4YqcNV0I/edit#gid=1959335211&range=A48:D106).

2. Every time a visit is created or updated in CommCare, determine which message(s) need to be scheduled. Schedule new messages and reschedule existing messages. 

For the sample case for [visit update](https://github.com/OpenFn/Miracle-Feet/blob/master/sample_data/newVisit_ReminderMessage.json) we have the following potential triggers in `state`:

`"send_sms": "on"`
`"sms_opt_in": "yes"`
 `"next_visit_date": "2021-05-18"`

Based on [alerts and their triggers](https://docs.google.com/spreadsheets/d/1quhQJgQkVRC8oObDzkwgnnm-Rov5BGOW85I4YqcNV0I/edit#gid=542770724), `alert 17` needs to be scheduled to be sent on `2021-05-16`.

Also from `state`:
`"language_code": ""`
the language of the messages has to be the default English.

That exact message that needs to be scheduled in this scenario is [here](https://docs.google.com/spreadsheets/d/1quhQJgQkVRC8oObDzkwgnnm-Rov5BGOW85I4YqcNV0I/edit#gid=1959335211&range=A112:D112).

## Toggl
https://track.toggl.com/3335160/projects/165487722/team

**Comment by aleksa-krolls** [comment]
@lakhassane whenever ready to start on this one, @ritazagoni is happy to walk you through it... but this is the same project you were looped in with Infobip.

**Comment by aleksa-krolls** [comment]
@lakhassane Whenever you get to start on this today, please advise on timelines for completion so that we can plan testing time accordingly. 
cc @daissatou @ritazagoni

**Comment by taylordowns2000** [comment]
Bumped estimate from 3 to 8 hours—spent the afternoon yesterday defining technical specifications for the bulk Id requirement _before_ any work on the expression could begin.

**Comment by aleksa-krolls** [comment]
@lakhassane status update please? We need to plan our testing and I need to prioritize accordingly, so please send a quick update every day - even if on slack cc @taylordowns2000 @ritazagoni

**Comment by lakhassane** [comment]
@aleksa-krolls still on it. Prioritinz with the others issue. I was stuck on the API side and with the `bulkId` specifications. 

Would you mind a meet chat for questions on it please? I think it will help better.

**Comment by aleksa-krolls** [comment]
@lakhassane @taylordowns2000 @ritazagoni per Infobip's email, it sounds like they were able to replicate the API issue with fetching via `bulkId`. Keep you posted... but a good reminder to ask for help! We often can reach out directly to tool providers when things aren't working as expected and make them do some of the troubleshooting work :)

**Comment by aleksa-krolls** [comment]
@lakhassane See below email from Shayne at Infobip. When scheduling SMSs... were we using a _future_ date when testing? When you have a chance, can you please try this out and let me know if it unblocks you to continue work on this request. 

_Could you try actually setting the time at a future date? I ran a few API calls right now and it seems to make the difference. So push the date to even December, and it should be good. Posted a few screenshots below. Do let me know what you see when you make these adjustments._

![image.png](https://images.zenhubusercontent.com/5d691cd53295e300013f1cbd/f9689690-b039-496b-b203-506cc8bcce87)
![image.png](https://images.zenhubusercontent.com/5d691cd53295e300013f1cbd/d97f290c-0c93-484d-ad9a-c5d2302fe2d9)
![image.png](https://images.zenhubusercontent.com/5d691cd53295e300013f1cbd/af44635d-fd9d-4806-8ace-560cc16d6262)

**Comment by lakhassane** [comment]
@aleksa-krolls tests are successful. I made a sendAt far on december and I am able to get it and even reschedule.
I also tested using special chars as `-` and it's working. So we might be able to keep going on this.

**Comment by aleksa-krolls** [comment]
@lakhassane Okay great, then please keep going. I will update state to show future dates. 

If any blockers, let me and @ritazagoni know so we can help you work through these.

**Comment by taylordowns2000** [comment]
@lakhassane , apologies if you've already sent this in another channel, but I want to make sure that this is ready for @aleksa-krolls and @ritazagoni when they get to work tomorrow: *we've got two options for how to build & maintain the sms text mapping*.

That big old translation table in google sheets needs to be represented as a JSON object. It's got message `sequences` > `messages` > `languages` with `parameters`.

In option 1, MiracleFeet must specify the _same_ `data` arguments for each translation of the text message and they'll have a much easier time maintaining this very large json object. In option 2 (which is how the excel spreadsheet is set up) they use different arguments for each language translation (so, the same message uses `firstName` and `date` in english, but `firstName`, `date`, and `gender` in French, for example) and they'll need to re-specify the parameters for each language in each message. The resulting JSON representation of that excel sheet is much bigger, but might be necessary. Do you know if we should go with option 1 or option 2?

## option 1
```json
{
  "patientReg": [
    {
      "EN": ["Hi, ", ". Your visit is on ", ". Please don't be late, ", "!"],
      "FR": [
        "Salut, ",
        ". votre reunion est a ",
        ". Soyez a bonne heure, ",
        "!"
      ],
      "delay": 0,
      "data": ["first_name", "nested.visitDate", "first_name"]
    },
    {
      "EN": [
        "Something else, ",
        ". I'll call you at ",
        ". Please pick up the phone!"
      ],
      "FR": ["Un autre, ", ". Je vais appeler a ", ". D'accord?"],
      "delay": 3200,
      "data": ["first_name", "nested.callTime"]
    }
  ]
}
```

## option 2
```json
{
  "patientReg": [
    {
      "EN": {
        "text": [
          "Hi, ",
          ". Your visit is on ",
          ". Please don't be late, ",
          "!"
        ],
        "data": ["first_name", "nested.visitDate", "first_name"]
      },
      "FR": {
        "text": ["Salut, ", ". votre reunion est a ", ". Soyez a bonne heure!"],
        "data": ["first_name", "nested.visitDate"]
      },
      "delay": 0
    },
    {
      "EN": {
        "text": [
          "Something else, ",
          ". I'll call you at ",
          ". Please pick up the phone!"
        ],
        "data": ["first_name", "nested.callTime"]
      },
      "FR": {
        "text": ["Un autre, ", ". Je vais appeler a ", ". D'accord?"],
        "data": ["first_name", "nested.callTime"]
      },
      "delay": 3200
    }
  ]
}
```

More on what's happening here: we'll determine which sequence `type` to schedule based on something in `state.data`. That's how we pick out `state.static.smsTranslations[type]`. From there, we get an array of, say, 12 messages to send over 30 days. For `each` message in that array, we'll pick out the right language, use `dataValue('first_name')` or whatever datavalue they need to create a message string, and schedule that message to be sent `now() + delay` time from now.

**Comment by aleksa-krolls** [comment]
@taylordowns2000 I honestly think both are tricky to manage and don't see a major different between the two. 
Not sure what @ritazagoni thinks, but is it possible to do something like the below? Otherwise I worry it's too hard to know where the dynamic fields fit into the sentence, and I can see the client messing this up. 
```
{
  "patientReg": [
    {
      "EN": {
        "text": `Hi, ${first_name}, your next visit is...`,
        "data": ["first_name", "nested.visitDate", "first_name"]
      },
      "FR": {
        "text": [`Hi, ${first_name}...`,
```
@lakhassane Also can you please provide an overall status update? Wondering when we can reasonably plan to test.

**Comment by taylordowns2000** [comment]
We could, but it would require regex. They would need to manually make sure
to adhere to a specific pattern and they messed it up there wouldn’t be a
simple way to detect it.

Mamadou, let’s discuss in the morning.

**Comment by aleksa-krolls** [comment]
Chatted with @ritazagoni ... let's go with Option 1 then for now... but if there's any middle ground approach that would make it easier on the client admins, let us know!

**Comment by taylordowns2000** [comment]
@ritazagoni , @aleksa-krolls , thanks for the help here today.

Status: 
1.	Nothing needed from services right now.
2.	Mamadou has the job code working. Tomorrow morning we will attempt to create message and alert schedules data from that spreadsheet that Rita finished up today. I expect that will take the two of us about 2hrs. It could be possible to test around noon utc. You could test smaller parts of the solution before then. 
3.	no blockers at the moment!

**Comment by aleksa-krolls** [comment]
@taylordowns2000 thanks a ton for this status update. 

@ritazagoni let's plan to get @daissatou involved with testing tomorrow!

**Comment by aleksa-krolls** [comment]
@taylordowns2000 @lakhassane when the job is ready to test, can you please configure it in prod here and advise on what configuration should look like? 
https://www.openfn.org/projects/419/jobs/3830

@ritazagoni Check out the trigger on that job, we will need to verify the CommCare values we can expect for state... e.g., I'm wondering if it should be `"sms_opt_in": "1"` (instead of "yes") - see this msg: https://www.openfn.org/projects/419/messages/10269882

**Comment by ritazagoni** [comment]
> @ritazagoni Check out the trigger on that job, we will need to verify the CommCare values we can expect for state... e.g., I'm wondering if it should be `"sms_opt_in": "1"` (instead of "yes") - see this msg: https://www.openfn.org/projects/419/messages/10269882

Right! Updated `"sms_opt_in": "1"` in https://docs.google.com/spreadsheets/d/1quhQJgQkVRC8oObDzkwgnnm-Rov5BGOW85I4YqcNV0I/edit#gid=262234774 fyi @lakhassane

**Comment by taylordowns2000** [comment]
@ritazagoni , I know you've already been in touch with @lakhassane , but I wanted to document that this is now ready for testing :-)

**Comment by ritazagoni** [comment]
@lakhassane we had a failed run https://openfn.org/projects/419/runs/10750769 on message https://openfn.org/projects/419/messages/10274332.

**Comment by lakhassane** [comment]
@ritazagoni @aleksa-krolls I think we hit another blocker in here. There are some criteria that are exactly the same but required different `bulkId`. Could you please help in a way to make the difference please?

### LINE 24

**Comment by aleksa-krolls** [comment]
@lakhassane So every 1 SMS that needs to be sent has a `bulkId` (1 `bulkId` for every 1 SMS alert). 
Sometimes the _same_ criteria might trigger multiple SMSs to be sent (1 trigger to many SMSs)... therefore there are multiple `bulkId` created via 1 triggering action. 

Real-world scenario: Patient comes into a client and enrolls in the `bracing_day` treatment --> they are then subscribed to a series of SMSs all related to bracing. They will receive multiple SMSs over a few days (each SMS will have its own `bulkId` and own message template). 

Does that help to explain? Or is there still a question here? Ping me on Slack if helpful to briefly discuss.

**Comment by lakhassane** [comment]
@ritazagoni @aleksa-krolls we find out that on different occassions the data we need to populate the sms content (`guardian_phone_number`, `guardian_first_name`,`patient_first_name`, ...) are not under the same object in the message received. 

For registration case type we can have it under `update`
```js
"update": {
          ...
          "phone_number_has_8digits": "no",
          "patient_first_name": "Rita Zagoni",
          "age_at_registration_months": "2",
          "registration_date": "2020-11-17", // needed
          "show_diagnosis_in_registration_form": "",
          "patient_last_name": "Zagoni",
          "guardian1_phone1_full": "25555555", // needed
          "clinic_name": "Shirati Hospital",
          "guardian1_first_name": "Test Guardian", // needed
          ...
        }
      },
```

For other case tests (`bracing_day`), we have the data we need it under `case_properties`
```js
"case_properties": {
          "case_name": "Rita Zagoni (TAN-KEN01-0075-01) 2020-11-17", 
          "clinic_code": "guinea_test_clinic",
          "clinic_name": "Ndanda Hospital", // needed
          "patient_dob": "2019-08-31",
          "patient_first_name": "Rita", // needd
          "patient_last_name": "Zagoni",
          "patient_name": "Rita Zagoni",
          "patient_original_id": null,
          "test_data": "0"
          ...
        },
```
Note that in the second case type, `guardian1_phone1_full`, used to send sms does not exist. 
There is no way to know beforehand where the data would be in the payload or how to access it.

Thinking out loud: Have we also considered splitting those case type in different jobs? They all have different `triggering criteria` (that could be used as trigger in prod) and a specific job would focus only on building the `bulkId` for a specific case type and sending/scheduling sms for that type.
job 1: smsRegistration.js -- trigger: `sms_opt_in = '1' AND send_sms = 'on'` 
job 2: bracingDay.js -- trigger: `sms_opt_in = '1' AND send_sms = 'on' AND treatment='bracing_day`
job 3: ..... 
@taylordowns2000 what's your thought on this.

**Comment by aleksa-krolls** [comment]
@lakhassane Sounds like we may need to consider 2 jobs here as well as provide updated mapping specifications for the different merge fields. Please pause development until we can analyze state further and provide updated specifications. 

Quick questions... 
1. Is everything at least working on the Infobip side? (I want to make sure we have no questions for this team.)
2. Where can I look to see what L3 of the job looks like? Is it [`content_output.json`](https://github.com/OpenFn/Miracle-Feet/blob/master/mappings/content-output.json)? 
3. What is the best way to provide updated mapping specifications for the merge fields in the SMS templates? 
For example...
if we need to specify a new path for `guardian1_first_name` for the below SMS, should we do this in-line in [column N](https://docs.google.com/spreadsheets/d/1quhQJgQkVRC8oObDzkwgnnm-Rov5BGOW85I4YqcNV0I/edit#gid=262234774)? Or should we create another column? 
'Hi {case.guardian1_first_name} this is from the clubfoot clinic at {case.clinic_name}. We will send you important information and reminders regarding {case.patient_first_name}'s treatment.'

**Comment by lakhassane** [comment]
@aleksa-krolls 
1. Yes everything is working api-side (infobip)
2. Exact. `content-output.json` will give it to you.
3.`Column N` will be more straightforward for development but a new column build only for path will be more clean so that we keep `live data` out of `text data`.

**Comment by aleksa-krolls** [comment]
@ritazagoni see `3` for how we can build a map for the SMS merge fields...
`live data/ path in state`: `text data/ merge field`
`case.update.guardian1.first_name`: `case.guardian1_first_name`

**Comment by taylordowns2000** [comment]
@aleksa-krolls and @ritazagoni , Mamadou and I have full x-ray vision on the job and mapping right now. We've made some readability and logging changes but are quite happy with our ability to deliver on business logic laid out in line 23-36. (See below.)

I'm going to shift this to review/QA now. From our side, we need to know:

1. What "from" phone number we should send these messages from.
2. How to update the logic (see below) which chooses which messages to send.
3. And _maybe_ talk through cases where the same bulkId key must be shared between messages. (Something like: `bracing_night: { '09D - Bracing Night Campaign Y1': 'bracing_night-' }, bracing_night: { '09D - Bracing Night Campaign Y2': 'bracing_night-' }`.) This isn't necessarily a critical discussion point, but it's possible that we could strip away some of the complexity of the job (keeping both a `bulkIds` array _and_ a `activeConditions` array feels clunky) if we better understood these edge cases.

Let us know when you'd like to talk and we can practice the requirements-to-dev handover as we shift this back to the backlog!

```js
  if (sms_opt_in === '1' && send_sms === 'on') {
    const { treatment } = calcs.sms;
    if (treatment === '') {
      bulkIds.push(treatmentMap['registration']);
      bulkIds.push(treatmentMap['treatmentIntro']);
      activeConditions.push(
        '01 - Patient Registration',
        '02 - Treatment Introduction'
      );
    } else {
      bulkIds.push(treatmentMap[treatment]);
      activeConditions.push('07 - Bracing All Day Campaign'); // replace with instruction to get key
    }
  }
```

**Comment by aleksa-krolls** [comment]
@taylordowns2000 @lakhassane We have updated specs coming your way. To address your latest feedback...
1. `from` sender phone numbers - @ritazagoni will provide
2. We have the new diagram to dictate the sending logic
3. `bulkId` will be unique for every alert (see column `M` in the master mapping document - e.g., `bracing_night_campaign-2...`, `bracing_night_campaign-3...`). We assume every 1 Message will have 1 `bulkId`... the bulkId key will always 

 @ritazagoni will submit a new issue shortly to maintain a record of this original spec.

**Comment by aleksa-krolls** [comment]
Updated issue specs and feedback are now being tracked on #32
